src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,525 @@ impl Display for SnapKey {
+                    snap_size += key.len();
+                    snap_size += value.len();
+                    snap_key_cnt += 1;
+                    try!(self.encode_compact_bytes(key));
+                    try!(self.encode_compact_bytes(value));
+                    Ok(true)
+                }));
+                // use an empty byte array to indicate that cf reaches an end.
+                box_try!(self.encode_compact_bytes(b""));
+            }
+            // use an empty byte array to indicate that kvpair reaches an end.
+            box_try!(self.encode_compact_bytes(b""));
+            try!(self.save_with_checksum());
+
+            info!("[region {}] scan snapshot, size {}, key count {}, takes {:?}",
+                  region.get_id(),
+                  snap_size,
+                  snap_key_cnt,
+                  t.elapsed());
+            Ok(())
+        }
+
+        fn total_size(&self) -> io::Result<u64> {
+            let meta = try!(self.meta());
+            Ok(meta.len())
+        }
+    }
+
+    impl Drop for SnapFile {
+        fn drop(&mut self) {
+            if let Some((_, path)) = self.tmp_file.take() {
+                debug!("deleting {}", path);
+                if let Err(e) = fs::remove_file(&path) {
+                    warn!("failed to delete temporary file {}: {:?}", path, e);
                 }
             }
-            self.res = Some(try!(self.reader.read_u32::<BigEndian>()));
         }
-        if self.res.unwrap() != self.digest.sum32() {
-            let msg = format!("crc not correct: {} != {}",
-                              self.res.unwrap(),
-                              self.digest.sum32());
-            return Err(io::Error::new(ErrorKind::InvalidData, msg));
+    }
+
+    /// A reader that calculate checksum and verify it without read
+    /// it from the beginning again.
+    pub struct SnapValidationReader {
+        reader: File,
+        digest: Digest,
+        left: usize,
+        res: Option<u32>,
+    }
+
+    impl SnapValidationReader {
+        /// Open the snap file located at specified path.
+        fn open<P: AsRef<Path>>(path: P) -> io::Result<SnapValidationReader> {
+            let reader = try!(File::open(path.as_ref()));
+            let digest = Digest::new(crc32::IEEE);
+            let len = try!(reader.metadata()).len();
+            if len < CRC32_BYTES_COUNT as u64 {
+                return Err(io::Error::new(ErrorKind::InvalidInput,
+                                          format!("file length {} < {}", len, CRC32_BYTES_COUNT)));
+            }
+            let left = len as usize - CRC32_BYTES_COUNT;
+            Ok(SnapValidationReader {
+                reader: reader,
+                digest: digest,
+                left: left,
+                res: None,
+            })
+        }
+
+        /// Validate the file
+        ///
+        /// If the reader will be consumed after calling this method, no further data can be
```

Comment:
This `validate` is called after all key values is read and written to RocksDB. So it's fine there's a restriction as this line of comment mentions.